optuna/_gp/acqf.py

```diff
@@ -22,6 +24,40 @@
     torch = _LazyImport("torch")
 
 
+def _sample_from_normal_sobol(dim: int, n_samples: int, seed: int | None) -> torch.Tensor:
+    # cf. https://github.com/pytorch/botorch/blob/v0.13.0/botorch/sampling/qmc.py#L26
+    sobol_engine = torch.quasirandom.SobolEngine(
+        dimension=dim, scramble=True, seed=seed
+    )  # type: ignore
+    # The Sobol sequence in [-1, 1].
+    samples = 2.0 * (sobol_engine.draw(n_samples, dtype=torch.float64) - 0.5)
+    # Inverse transform to standard normal (values to close to 0/1 result in inf values).
+    return torch.erfinv(samples) * float(np.sqrt(2))
+
+
+def logehvi(
+    Y_post: torch.Tensor,  # (..., n_qmc_samples, n_objectives)
+    non_dominated_box_lower_bounds: torch.Tensor,  # (n_boxes, n_objectives)
+    non_dominated_box_upper_bounds: torch.Tensor,  # (n_boxes, n_objectives)
+) -> torch.Tensor:  # (..., )
+    log_n_qmc_samples = float(np.log(Y_post.shape[-2]))
+    # This function calculates Eq. (1) of https://arxiv.org/abs/2006.05078.
+    # TODO(nabenabe): Adapt to Eq. (3) when we support batch optimization.
+    # TODO(nabenabe): Make the calculation here more numerically stable.
+    # cf. https://arxiv.org/abs/2310.20708
+    # Check the implementations here:
+    # https://github.com/pytorch/botorch/blob/v0.13.0/botorch/utils/safe_math.py
+    # https://github.com/pytorch/botorch/blob/v0.13.0/botorch/acquisition/multi_objective/logei.py#L146-L266
+    _EPS = torch.tensor(1e-12, dtype=torch.float64)  # NOTE(nabenabe): grad becomes nan when EPS=0.
+    diff = torch.maximum(
+        _EPS,
+        torch.minimum(Y_post[..., torch.newaxis, :], non_dominated_box_upper_bounds)
+        - non_dominated_box_lower_bounds,
+    )
+    log_hvi_vals = torch.special.logsumexp(diff.log().sum(dim=-1), dim=-1)
+    return -log_n_qmc_samples + torch.special.logsumexp(log_hvi_vals, dim=-1)
```

Comment:
This can be simply written as follows:
```suggestion
    return -log_n_qmc_samples + torch.special.logsumexp(diff.log().sum(dim=-1), dim=(-2, -1))
```